File: backend/api/routes/bankroll.py
"""Bankroll endpoints — balance, deposit/withdraw, place bets, settle, ROI."""

from fastapi import APIRouter, Query
from sqlalchemy import select

from api.deps import BankrollDep, DbSession
from api.schemas.bankroll import (
    BalanceResponse,
    BetResponse,
    DepositRequest,
    HistoryItemResponse,
    HistoryResponse,
    PlaceBetRequest,
    RoiResponse,
    SettleBetRequest,
    SettleBetResponse,
    SnapshotResponse,
    WithdrawRequest,
)
from db.enums import BetOutcome
from db.models import BankrollSnapshot


router = APIRouter(prefix="/api/bankroll", tags=["bankroll"])


@router.get("", response_model=BalanceResponse)
def get_balance(tracker: BankrollDep) -> BalanceResponse:
    """Current bankroll state — total, committed in pending bets, available."""
    return BalanceResponse(
        current=tracker.get_current_balance(),
        available=tracker.get_available_balance(),
        pending_commitment=tracker.get_pending_commitment(),
    )


@router.post("/deposit", response_model=SnapshotResponse, status_code=201)
def deposit(body: DepositRequest, tracker: BankrollDep) -> SnapshotResponse:
    """Add funds to the bankroll."""
    snapshot = tracker.deposit(body.amount)
    return SnapshotResponse.model_validate(snapshot)


@router.post("/withdraw", response_model=SnapshotResponse, status_code=201)
def withdraw(body: WithdrawRequest, tracker: BankrollDep) -> SnapshotResponse:
    """Remove funds from the bankroll. Fails if committed capital would go negative."""
    snapshot = tracker.withdraw(body.amount)
    return SnapshotResponse.model_validate(snapshot)


@router.post("/bets", response_model=BetResponse, status_code=201)
def place_bet(body: PlaceBetRequest, tracker: BankrollDep) -> BetResponse:
    """Place a bet against a prediction. Locks stake until settle."""
    bet = tracker.place_bet(body.prediction_id, body.quota, body.stake)
    return BetResponse.model_validate(bet)


@router.post("/bets/{bet_id}/settle", response_model=SettleBetResponse)
def settle_bet(
    bet_id: int, body: SettleBetRequest, tracker: BankrollDep,
) -> SettleBetResponse:
    """Settle a pending bet as won, lost, or void. Updates bankroll accordingly."""
    bet, snapshot = tracker.settle_bet(bet_id, BetOutcome(body.outcome))
    return SettleBetResponse(
        bet=BetResponse.model_validate(bet),
        snapshot=SnapshotResponse.model_validate(snapshot) if snapshot else None,
    )


@router.get("/roi", response_model=RoiResponse)
def get_roi(tracker: BankrollDep) -> RoiResponse:
    """Win rate + net profit + ROI across all settled bets."""
    report = tracker.compute_roi()
    return RoiResponse(**report.__dict__)


@router.get("/history", response_model=HistoryResponse)
def get_history(db: DbSession, limit: int = Query(20, ge=1, le=200)) -> HistoryResponse:
    """Recent bankroll movements (deposits, withdrawals, bet outcomes)."""
    rows = db.execute(
        select(BankrollSnapshot)
        .order_by(BankrollSnapshot.created_at.desc())
        .limit(limit)
    ).scalars().all()
    return HistoryResponse(
        items=[HistoryItemResponse.model_validate(r) for r in rows],
        total=len(rows),
    )

File: mobile/database/bankroll.js
// Port of backend/core/bankroll/tracker.py — same append-only design.
// Balance = last snapshot's balance; available = balance - pending bets.
// Deposits / withdrawals / settle outcomes all INSERT new snapshots.

function lastBalance(db) {
  const row = db.getFirstSync(
    `SELECT balance FROM bankroll_snapshots
     ORDER BY created_at DESC, id DESC LIMIT 1`,
  );
  return row ? row.balance : 0.0;
}

function pendingCommitment(db) {
  const row = db.getFirstSync(
    `SELECT COALESCE(SUM(stake_amount), 0.0) AS s FROM bets
     WHERE outcome = 'pending'`,
  );
  return row.s || 0.0;
}

export function getBalance(db) {
  const current = lastBalance(db);
  const pending = pendingCommitment(db);
  return {
    current,
    available: current - pending,
    pending_commitment: pending,
  };
}

export function getHistory(db, limit = 20) {
  const items = db.getAllSync(
    `SELECT id, balance, change_amount, reason, related_bet_id, created_at
     FROM bankroll_snapshots
     ORDER BY created_at DESC, id DESC
     LIMIT ?`,
    [limit],
  );
  return { items, total: items.length };
}

function insertSnapshot(db, { changeAmount, reason, relatedBetId = null }) {
  const newBalance = lastBalance(db) + changeAmount;
  const createdAt = new Date().toISOString();
  const result = db.runSync(
    `INSERT INTO bankroll_snapshots
     (balance, change_amount, reason, related_bet_id, created_at)
     VALUES (?, ?, ?, ?, ?)`,
    [newBalance, changeAmount, reason, relatedBetId, createdAt],
  );
  return {
    id: result.lastInsertRowId,
    balance: newBalance,
    change_amount: changeAmount,
    reason,
    related_bet_id: relatedBetId,
    created_at: createdAt,
  };
}

export function deposit(db, amount) {
  if (!(amount > 0)) {
    throw new Error('El monto a depositar debe ser mayor a 0.');
  }
  return insertSnapshot(db, { changeAmount: amount, reason: 'deposit' });
}

export function withdraw(db, amount) {
  if (!(amount > 0)) {
    throw new Error('El monto a retirar debe ser mayor a 0.');
  }
  const { available } = getBalance(db);
  if (amount > available) {
    throw new Error(
      `No alcanza el disponible ($${available.toFixed(2)}) para retirar $${amount.toFixed(2)}.`,
    );
  }
  return insertSnapshot(db, { changeAmount: -amount, reason: 'withdrawal' });
}

// Internal — called by bets.settleBet after marking the bet outcome.
export function recordSettlement(db, { betId, outcome, stake, quota }) {
  if (outcome === 'won') {
    const profit = (quota - 1.0) * stake;
    return insertSnapshot(db, {
      changeAmount: profit,
      reason: 'bet_won',
      relatedBetId: betId,
    });
  }
  if (outcome === 'lost') {
    return insertSnapshot(db, {
      changeAmount: -stake,
      reason: 'bet_lost',
      relatedBetId: betId,
    });
  }
  // 'void' — full refund, no net balance change → no snapshot.
  return null;
}
